### PR TITLE
Add quick fix for S1172 ('no-unused-function-argument')

### LINKED ISSUE #3033: Add quick fix for S1172: Unused function parameters should be removed
SonarJS rule.

### PR DESCRIPTION
**Pull request by yassin-kammoun-sonarsource**
Fixes #3033

### CODE REVIEW

File: eslint-bridge/src/rules/no-unused-function-argument.ts [hunk 1]
```diff
@@ -80,11 +83,49 @@ function reportUnusedArgument(
...
+  const suggestions: Rule.SuggestionReportDescriptor[] = [
+    {
+      messageId: 'suggestRenameParameter',
+      data: {
+        param: paramVariable.name,
+      },
+      fix: fixer => fixer.replaceText(paramIdentifier, `_${paramVariable.name}`),
+    },
+  ];
+  const funct = paramVariable.defs[0].node as TSESTree.FunctionLike;
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-unused-function-argument.ts:103]: why such name for variable?

File: eslint-bridge/src/rules/no-unused-function-argument.ts [hunk 2]
```diff
@@ -80,11 +83,49 @@ function reportUnusedArgument(
...
 
+function getSuggestions(paramVariable: Scope.Variable) {
+  const paramIdentifier = paramVariable.identifiers[0];
+  const suggestions: Rule.SuggestionReportDescriptor[] = [
+    {
+      messageId: 'suggestRenameParameter',
+      data: {
+        param: paramVariable.name,
+      },
+      fix: fixer => fixer.replaceText(paramIdentifier, `_${paramVariable.name}`),
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-unused-function-argument.ts:100]: may be using `insertTextBefore` will be more efficient?

File: eslint-bridge/src/rules/no-unused-function-argument.ts (hunk 1 above)
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-unused-function-argument.ts:103]: why such name for variable?

File: eslint-bridge/tests/rules/no-unused-function-argument.test.ts [hunk 3]
```diff
@@ -196,5 +196,54 @@ ruleTester.run('Unused function parameters should be removed', rule, {
             }`,
       errors: 1,
     },
+    {
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/no-unused-function-argument.test.ts:199]: can we add unit test with parameter type to verify that type is also removed?

File: eslint-bridge/src/rules/no-unused-function-argument.ts [hunk 4]
```diff
@@ -80,11 +83,49 @@ function reportUnusedArgument(
...
+  if ((paramIdentifier as TSESTree.Node).parent === funct) {
+    suggestions.push({
+      messageId: 'suggestRemoveParameter',
+      data: {
+        param: paramVariable.name,
+      },
+      fix: fixer => {
+        const paramIndex = funct.params.indexOf(paramIdentifier as TSESTree.Parameter);
+        if (funct.params.length === 1) {
+          return fixer.remove(funct.params[paramIndex] as estree.Node);
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-unused-function-argument.ts:113]: what will happen with trailing comma?

File: eslint-bridge/src/rules/no-unused-function-argument.ts [hunk 5]
```diff
@@ -80,11 +83,49 @@ function reportUnusedArgument(
...
+        } else if (funct.params.length - 1 === paramIndex) {
+          const [, start] = funct.params[paramIndex - 1].range;
+          const [, end] = funct.params[paramIndex].range;
+          return fixer.removeRange([start, end]);
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-unused-function-argument.ts:117]: this should be fine for trailing comma but can we add a test for it?

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: eslint-bridge/src/rules/no-unused-function-argument.ts (hunk 1 above)
Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/rules/no-unused-function-argument.ts:103]: `funct` for `function`, but the latter is a keyword.

Comment (vilchik-elena) [reply eslint-bridge/src/rules/no-unused-function-argument.ts:103]: ah, indeed:)) it's just usually we use `fun` or `func`. I have never seen `funct`

File: eslint-bridge/src/rules/no-unused-function-argument.ts [hunk 6]
```diff
@@ -80,11 +88,90 @@ function reportUnusedArgument(
...
+          case 'ArrowFunctionExpression': {
+            const param = func.params[paramIndex] as estree.Node;
+            const token = context.getSourceCode().getTokenAfter(param);
+            if (token?.value === ')') {
+              return fixer.remove(param);
+            } else {
+              return fixer.replaceText(param, '()');
+            }
+          }
+          case 'FunctionDeclaration': {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-unused-function-argument.ts:139]: this case and next one are very similar, can't we generalize the logic?

File: eslint-bridge/src/rules/no-unused-function-argument.ts [hunk 7]
```diff
@@ -23,11 +23,19 @@ import { Rule, Scope } from 'eslint';
...
+  | TSESTree.FunctionDeclaration
+  | TSESTree.FunctionExpression;
+
 export const rule: Rule.RuleModule = {
   meta: {
+    hasSuggestions: true,
     messages: {
       removeOrRenameParameter:
         'Remove the unused function parameter "{{param}}" or rename it to "_{{param}}" to make intention explicit.',
+      suggestRemoveParameter: 'Remove {{param}} (beware of call sites)',
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-unused-function-argument.ts:37]: I think we should wrap parameter names into quotes.

File: eslint-bridge/src/rules/no-unused-function-argument.ts [hunk 8]
```diff
@@ -80,11 +88,90 @@ function reportUnusedArgument(
...
+    data: {
+      param: paramVariable.name,
+    },
+    fix: fixer => {
+      const paramIndex = func.params.indexOf(paramIdentifier as TSESTree.Parameter);
+      if (func.params.length === 1) {
+        switch (func.type) {
+          case 'ArrowFunctionExpression': {
+            const param = func.params[paramIndex] as estree.Node;
+            const token = context.getSourceCode().getTokenAfter(param);
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-unused-function-argument.ts:132]: theoretically there could be comma (`(a,) => foo()`)

File: eslint-bridge/tests/rules/no-unused-function-argument.test.ts [hunk 9]
```diff
@@ -196,5 +196,180 @@ ruleTester.run('Unused function parameters should be removed', rule, {
...
+      code: `function fun(a, b, c, d, {e}, f, g) { b = d = f; }`,
+      errors: [
+        {
+          suggestions: [
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/no-unused-function-argument.test.ts:203]: I think you don't have to assert all suggestions and their messages. Assert only what matters for the case, otherwise it's hard to follow

**Review by vilchik-elena — CHANGES_REQUESTED** [review]
And same thing as with another PR, I think we should keep the trailing comma style

File: eslint-bridge/src/rules/no-unused-function-argument.ts [hunk 10]
```diff
@@ -80,11 +88,83 @@ function reportUnusedArgument(
...
+      param: paramVariable.name,
+    },
+    fix: fixer => {
+      const paramIndex = func.params.indexOf(paramIdentifier as TSESTree.Parameter);
+      const param = func.params[paramIndex] as estree.Node;
+      if (func.params.length === 1) {
+        const openingParenthesis = context
+          .getSourceCode()
+          .getTokenBefore(param, token => token.value === '(');
+        const closingParenthesis = context
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-unused-function-argument.ts:133]: this will not work for `a => foo()` :(

File: eslint-bridge/src/rules/no-unused-function-argument.ts [hunk 11]
```diff
@@ -80,11 +88,83 @@ function reportUnusedArgument(
...
+          .getSourceCode()
+          .getTokenAfter(param, token => token.value === ')');
+        let [start, end] = param.range!;
+        if (openingParenthesis) {
+          start = openingParenthesis.range[0];
+        }
+        if (closingParenthesis) {
+          end = closingParenthesis.range[1];
+        }
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-unused-function-argument.ts:142]:
```suggestion
        const openingParenthesis = context
          .getSourceCode()
          .getTokenBefore(param);
        const closingParenthesis = context
          .getSourceCode()
          .getTokenAfter(param, token => token.value === ')');
        let [start, end] = param.range!;
        if (openingParenthesis && openingParenthesis.value === '(') {
          start = openingParenthesis.range[0];
          end = closingParenthesis!.range[1];
        }
```

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: eslint-bridge/tests/rules/no-unused-function-argument.test.ts [hunk 12]
```diff
@@ -196,5 +196,163 @@ ruleTester.run('Unused function parameters should be removed', rule, {
...
+    },
+    {
+      // arrow with trailing
+      code: `( a, ) => {}`,
+      errors: [
+        {
+          suggestions: [
+            {
+              desc: 'Rename "a" to "_a"',
+              output: '( _a, ) => {}',
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/no-unused-function-argument.test.ts:347]: I see we can't remove this suggestion from assertion, but let's at least remove the messages to reduce some readablility noise

**Review by vilchik-elena — APPROVED** [review]
